src/components/basket/Basket.js

```diff
@@ -0,0 +1,164 @@
+import React from 'react';
+
+import dataApi from 'api/data';
+import BasketDetails from './BasketDetails';
+import BasketButton from './BasketButton';
+import Loader from 'components/common/Loader';
+
+class Basket extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      basketItems: [],
+      basketCountItems: 0,
+      basketTotalPrice: 0,
+      isBasketDetailHidden: true,
+      isLoading: false,
+      requestErrMsg: null,
+    };
+  }
+
+  componentDidMount() {
+    this._loadInitBasket();
+  }
+
+  componentDidUpdate() {
+    if (this.props.newDishItemToBasket && !this.state.isLoading) {
+      this.addDishToBasket(this.props.newDishItemToBasket);
+    }
+  }
+
+  _onLoadChange = (isLoading) => {
+    this.props.onChangeBasketBusy(isLoading);
+    this.setState({ isLoading: isLoading });
+  };
+
+  _loadInitBasket = async () => {
+    this._onLoadChange(true);
+    try {
+      const basket = await dataApi.get('basket');
+      const [calcBasketCountItems, calcBasketTotalPrice] = this._calcBasketInfo(
+        basket.data
+      );
+      this.setState({
+        basketItems: [...basket.data],
+        basketCountItems: calcBasketCountItems,
+        basketTotalPrice: calcBasketTotalPrice,
+      });
+    } catch (e) {
+      this.setState({ requestErrMsg: e.toString() });
+    } finally {
+      this._onLoadChange(false);
+    }
+  };
+
+  _calcBasketInfo = (basketItems) => {
+    const basketInfo = basketItems.reduce(
+      (prev, current) => {
+        prev[0] += current.count;
+        prev[1] += current.price;
+        return prev;
+      },
+      [0, 0]
+    );
+    return basketInfo;
+  };
+
+  toggleBasketDetail = () => {
+    this.setState({ isBasketDetailHidden: !this.state.isBasketDetailHidden });
+  };
+
+  addDishToBasket = async (newItem) => {
+    const existItemIdx = this.state.basketItems.findIndex(
+      (item) => item.id === newItem.id
+    );
+    const newBasketItems = [...this.state.basketItems];
+    this._onLoadChange(true);
+    try {
+      if (existItemIdx !== -1) {
+        newBasketItems[existItemIdx].count++;
+        newBasketItems[existItemIdx].price += newItem.price;
+        await dataApi.put(`basket/${newBasketItems[existItemIdx].id}`, {
+          ...newBasketItems[existItemIdx],
+        });
+      } else {
+        const results = {
+          id: newItem.id,
+          name: newItem.name,
+          price: newItem.price,
+          count: 1,
+        };
+        await dataApi.post('basket', results);
```

Comment:
Imagine the case when dataApi.put('basket') (line 81) completed successfully, but  dataApi.post('basket') not.
Then the data in json file will be updated, but not on the UI becase setting the data to the state is located after post API call